### PR TITLE
Added MinGW32 SEH crash and assert protection

### PR DESCRIPTION
**Pull request by clibequilibrium**
This PR adds crash protection on MinGW32 . Please note since __except is not supported on MinGW I had to use __try1 and __except1 . The only problem with my solution is that there is no way for me to get the right error code so I just return CR_OTHER. 

Also is returning -1 okay after try catch ? From my tests it introduced the correct behavior of rollback. Previously crash protection was not working at all for me. This PR tackles this issue. 

~~Please note while this PR adds crash protection it doesn't add assertation protection yet.~~

EDIT: the PR has been updated to fully support assert and crash protection in MingW32 environment

### CODE REVIEW

File: cr.h [hunk 1]
```diff
@@ -1138,15 +1168,27 @@ static int cr_plugin_main(cr_plugin &ctx, cr_op operation) {
...
         return -1;
     }
+#else
+    if (int sig = __builtin_setjmp(env)) {
+      ctx.version = ctx.last_working_version;
+      ctx.failure = cr_signal_to_failure(sig);
+      CR_LOG("1 FAILURE: %d (CR: %d)\n", sig, ctx.failure);
+      return -1;
+    } else {
+      auto p = (cr_internal *)ctx.p;
```
Comment (fungos) [inline cr.h:1184]: You're shadowing the p from the start of the function here, I think you can remove this line

File: cr.h [hunk 2]
```diff
@@ -1103,7 +1103,37 @@ static cr_plugin_main_func cr_so_symbol(so_handle handle) {
...
+#include <setjmp.h>
+#include <signal.h>
+
+static jmp_buf env;
+static void cr_signal_handler(int sig) {
+    __builtin_longjmp(env, 1);
+}
+
+static cr_failure cr_signal_to_failure(int sig) {
+  switch (sig) {
```
Comment (fungos) [inline cr.h:1116]: Overall this looks good to me, unfortunately our CI tests are broken due travis ending support and I did not migrate it to Github Actions, so I'll be going a bit blindly here. :crossed_fingers: 

Note that I would like you to keep the standard formatting in the file (4-spaces instead 2-spaces from your changes), so please fix that too.

**Review by fungos — CHANGES_REQUESTED** [review]

File: cr.h (hunk 1 above)
Comment (clibequilibrium) [reply cr.h]: @fungos  the line was removed, well spotted

File: cr.h (hunk 2 above)
Comment (clibequilibrium) [reply cr.h]: My bad, I did not applied clang formatting to my code block. 4 spaces were corrected.

**Review by fungos — APPROVED** [review]

**Comment by fungos** [comment]
Thanks for fixing this!

**Pull request merged**